sklearn/utils/extmath.py

```diff
@@ -707,6 +707,91 @@ def _safe_accumulator_op(op, x, *args, **kwargs):
     return result
 
 
+def _incremental_weighted_mean_and_var(X, sample_weight,
+                                       last_mean,
+                                       last_variance,
+                                       last_weight_sum):
+    """Calculate weighted mean and variance batch update
+
+    last_mean and last_variance are statistics computed at the last step
+    by the function. Both must be initialized to 0.0. In case no scaling
+    is required last_variance can be None. The mean is always required and
+    returned because necessary for the calculation of the variance.
+    last_weight sum is the sum of weights encountered until now.
+
+    Parameters
+    ----------
+    X : array-like, shape (n_samples, n_features)
+        Data to use for statistics update
+
+    sample_weight : array-like of shape (n_samples,)
+
+    last_mean : array-like of shape: (n_features,)
+
+    last_variance : array-like of shape: (n_features,)
+
+    last_weight_sum : array-like of shape (n_features,)
+
+    Returns
+    -------
+    updated_mean : array of shape (n_features,)
+
+    updated_variance : array of shape (n_features,)
+        If None, only mean is computed
+
+    updated_weight_sum : array of shape (n_features,)
+
+    Notes
+    -----
+    NaNs in X are ignored.
+
+    References
+    ----------
+    Tony Finch
+    "Incremental calculation of weighted mean and variance"
+    University of Cambridge Computing Service, February 2009
+
+    """
+    # last = stats until now
+    # new = the current increment
+    # updated = the aggregated stats
+
+    nan_mask = np.isnan(X)
+    sample_weight_T = np.reshape(sample_weight, (1, -1))
+    new_weight_sum = \
```

Comment:
you might just remind the reader that this has shape `(n_features,)` (to parallel last_weight_sum)